Native c10d_functional ops

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109738

_c10d_functional is a native port of the existing c10d_functional from
Python. _c10d_functional will supercede c10d_functional upon confirmation of
feature parity.

Reasons for transitioning c10d_functional ops to C++:
- Enable collective ops for AOT Inductor and native interpreters.
- Enable collective ops for multi-threaded native runtimes.
- Unify collective codegens between Python Inductor and AOT Inductor.

Difference from the Python c10d_functional:
- The namespace now included in-place variation of collectives calls, which
will be used by Inductor wrapper for de-functionalized collective calls.
- Process group resolution now only relies on tag. Ranks are useful for the
compiler to determine whether two collectives using different process groups
are fuse-able. However, they are not required for process group resolution.
- Process group resolution now doesn't upsert a process group.

Differential Revision: [D49479258](https://our.internmc.facebook.com/intern/diff/D49479258)

Commit:
Native c10d_functional

[ghstack-poisoned]

yifuwang commented:
@yifuwang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49479258).

SherlockNoMad commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -0,0 +1,280 @@
+    const at::Tensor& input,
+    int64_t groupSize) {
+  auto outputSize = input.sizes().vec();
+  outputSize[0] /= groupSize;
```
I wonder if you need to handle non-divisible case.

SherlockNoMad commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -0,0 +1,280 @@
+
+at::Tensor allocateAllGatherOutput(const at::Tensor& input, int64_t groupSize) {
+  auto outputSize = input.sizes().vec();
+  outputSize[0] *= groupSize;
```
is AllGather always expanding on the first dim ?

yifuwang replied:
Yes. This is the behavior for the current API.

SherlockNoMad commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -0,0 +1,280 @@
+    at::Tensor input,
+    const std::string& reduceOp,
+    const std::string& tag,
+    const std::vector<int64_t> ranks,
```
curious how `ranks` is used? 

I don't see this used anywhere.

yifuwang replied:
`ranks` is intended to be consumed by compilers. For example, it's possible to fuse two all_reduces issued with two groups of the same membership. However, compilers can't determine the legitimacy of such fusion in absence of membership information.

The Python user facing APIs also leverages the information for lazy sub-pg initialization (I don't think we should do the same in native mode though).

SherlockNoMad commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -0,0 +1,280 @@
+          c10::DispatchKey::CompositeExplicitAutograd, ::all_reduce));
+
+  m.def(
+      "all_reduce_(Tensor self, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor",
```
schema should be 
```
"all_reduce_(Tensor(a!) self, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor(a!)"
```

SherlockNoMad commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -0,0 +1,280 @@
+          c10::DispatchKey::CompositeExplicitAutograd, ::all_reduce_coalesced));
+
+  m.def(
+      "all_reduce_coalesced_(Tensor[] self, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor[]",
```
need to use `Tensor(a!)[] self `

SherlockNoMad commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -0,0 +1,280 @@
+          ::reduce_scatter_tensor_coalesced));
+
+  m.def(
+      "wait_tensor(Tensor self) -> Tensor",
```
This is an inplace op, right? 
If yes, it should use `Tensor(a!) self`

SherlockNoMad commented:
Need tests

Commit:
Update on "Native c10d_functional ops"


_c10d_functional is a native port of the existing c10d_functional from
Python. _c10d_functional will supercede c10d_functional upon confirmation of
feature parity.

Reasons for transitioning c10d_functional ops to C++:
- Enable collective ops for AOT Inductor and native interpreters.
- Enable collective ops for multi-threaded native runtimes.
- Unify collective codegens between Python Inductor and AOT Inductor.

Difference from the Python c10d_functional:
- The namespace now included in-place variation of collectives calls, which
will be used by Inductor wrapper for de-functionalized collective calls.
- Process group resolution now only relies on tag. Ranks are useful for the
compiler to determine whether two collectives using different process groups
are fuse-able. However, they are not required for process group resolution.
- Process group resolution now doesn't upsert a process group.

Differential Revision: [D49479258](https://our.internmc.facebook.com/intern/diff/D49479258)

[ghstack-poisoned]

yifuwang reviewed (commented):
Added tests.

Commit:
Update on "Native c10d_functional ops"


_c10d_functional is a native port of the existing c10d_functional from
Python. _c10d_functional will supercede c10d_functional upon confirmation of
feature parity.

Reasons for transitioning c10d_functional ops to C++:
- Enable collective ops for AOT Inductor and native interpreters.
- Enable collective ops for multi-threaded native runtimes.
- Unify collective codegens between Python Inductor and AOT Inductor.

Difference from the Python c10d_functional:
- The namespace now included in-place variation of collectives calls, which
will be used by Inductor wrapper for de-functionalized collective calls.
- Process group resolution now only relies on tag. Ranks are useful for the
compiler to determine whether two collectives using different process groups
are fuse-able. However, they are not required for process group resolution.
- Process group resolution now doesn't upsert a process group.

Differential Revision: [D49479258](https://our.internmc.facebook.com/intern/diff/D49479258)

[ghstack-poisoned]

yifuwang commented:
@yifuwang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49479258).

Commit:
Update on "Native c10d_functional ops"


_c10d_functional is a native port of the existing c10d_functional from
Python. _c10d_functional will supercede c10d_functional upon confirmation of
feature parity.

Reasons for transitioning c10d_functional ops to C++:
- Enable collective ops for AOT Inductor and native interpreters.
- Enable collective ops for multi-threaded native runtimes.
- Unify collective codegens between Python Inductor and AOT Inductor.

Difference from the Python c10d_functional:
- The namespace now included in-place variation of collectives calls, which
will be used by Inductor wrapper for de-functionalized collective calls.
- Process group resolution now only relies on tag. Ranks are useful for the
compiler to determine whether two collectives using different process groups
are fuse-able. However, they are not required for process group resolution.
- Process group resolution now doesn't upsert a process group.

Differential Revision: [D49479258](https://our.internmc.facebook.com/intern/diff/D49479258)

[ghstack-poisoned]

SherlockNoMad commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -0,0 +1,285 @@
+          c10::DispatchKey::CompositeExplicitAutograd, ::all_reduce));
+
+  m.def(
+      "all_reduce_(Tensor(a!) self, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor",
```
all need to tag output. 

see examples in native_functions.yaml
e.g.
`- func: normal_(Tensor(a!) self, float mean=0, float std=1, *, Generator? generator=None) -> Tensor(a!)`

SherlockNoMad replied:
also see https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/README.md

SherlockNoMad commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -0,0 +1,285 @@
+// - Process group resolution now doesn't upsert a process group.
+
+static thread_local std::unordered_map<void*, c10::intrusive_ptr<c10d::Work>>
+    pendingWork;
```
Also, pytorch c++ code uses snake_case names.

yifuwang commented on torch/csrc/distributed/c10d/ProcessGroup.cpp:
```diff
@@ -159,10 +159,36 @@ const std::string& ProcessGroup::getGroupName() const {
   return deviceTypeToBackend_.begin()->second->getGroupName();
 }
 
+static thread_local std::
```
Good catch. I initially implemented this purely as a backend for AOT Inductor and later pivoted it as a replacement for `c10d_functional`. You are right that autograd engine won't be able to resolve the pg correctly.

> If the idea of using a thread local is to support the multi-rank PG

I'm not sure if it's the same thing, but the motivation is to enable the one-thread-per-device concurrency model in native runtimes (let each thread be a rank).

yifuwang commented on torch/csrc/distributed/c10d/ProcessGroup.cpp:
```diff
@@ -159,10 +159,36 @@ const std::string& ProcessGroup::getGroupName() const {
+        groupRegistry;
+
 void ProcessGroup::setGroupName(const std::string& name) {
+  if (groupRegistry.find(name) != groupRegistry.end()) {
```
Sounds good. Let's move the entire mechanism to native.

yifuwang commented:
Superceded by https://github.com/pytorch/pytorch/pull/110570
